cpp/include/cudf/detail/reduction_functions.hpp

```diff
@@ -254,5 +254,140 @@ std::unique_ptr<scalar> nth_element(
+ * @param mr Device memory resource used to allocate the returned scalar's device memory
+ * @return Product as scalar of type `output_dtype`.
+ */
+std::unique_ptr<column> segmented_product(
+  column_view const& col,
+  column_view const& offsets,
+  data_type const output_dtype,
+  null_policy null_handling,
+  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
+
+/**
+ * @brief Compute minimum of each segment in input column
+ *
+ * If all elements in input segment are null, the segment result is null.
+ *
+ * @throw cudf::logic_error if input column type is convertible to `output_dtype`
+ *
+ * @param col Input column to compute minimum.
+ * @param offsets Indices to identify segment boundaries
+ * @param output_dtype Data type of return type and typecast elements of input column
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ * @param mr Device memory resource used to allocate the returned scalar's device memory
+ * @return Minimums of segments in type `output_dtype`.
+ */
+std::unique_ptr<column> segmented_min(
+  column_view const& col,
+  column_view const& offsets,
+  data_type const output_dtype,
+  null_policy null_handling,
+  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
+
+/**
+ * @brief Compute maximum of each segment in input column
+ *
+ * If all elements in input segment are null, the segment result is null.
+ *
+ * @throw cudf::logic_error if input column type is convertible to `output_dtype`
+ *
+ * @param col Input column to compute maximum.
+ * @param offsets Indices to identify segment boundaries
+ * @param output_dtype Data type of return type and typecast elements of input column
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ * @param mr Device memory resource used to allocate the returned scalar's device memory
+ * @return Maximums of segments in type `output_dtype`.
+ */
+std::unique_ptr<column> segmented_max(
+  column_view const& col,
+  column_view const& offsets,
+  data_type const output_dtype,
+  null_policy null_handling,
+  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
+
+/**
+ * @brief Compute any of the value in the segment is true when typecasted to bool
+ *
+ * If all elements in input segment are null, the segment result is null.
+ *
+ * @throw cudf::logic_error if input column type is not convertible to bool
+ * @throw cudf::logic_error if `output_dtype` is not bool
+ *
+ * @param col Input column to compute any_of.
+ * @param offsets Indices to identify segment boundaries
+ * @param output_dtype Data type of return type and typecast elements of input column
```

Comment:
Maybe? In `simple_segmented.cuh` this may help getting rid of `bool_resul_column_dispatcher`. I think the performance should be similar.

Reply:
I noticed `any` is using `min` to implement, that means the result is being casted to bool regardless. To really support any `output_dtype` you would need an extra casting step. So ignore what I said above.

Reply:
I looked at this with @isVoid. We have to keep this `output_type` parameter because it is exposed to the user API. That way, we can show an error if the user requests an "all" or "any" reduction with a type besides bool8. While it's a little clunky, it is probably better to error than to cast the resulting `bool8` column to some other type like `double` or `int32`. Users needing to cast the result can do so explicitly.